### PR TITLE
Convert Trace Serialization mechanism from SOAP to JSON

### PR DESCRIPTION
**Pull request by saintentropy**
### Purpose

Experiment with alternative to the SOAP Formatter utilized in trace and Element binding data.  This PR makes two primary changes related to trace:

Modify the public `TraceSupport` APIs to utilize `string` as the storage mechanism vs `ISerializable`.   These are the primary mechanism for implementing Trace within ZT implementation.  
- `ISerializable GetTraceData(string key)` -> `sting GetTraceData(string key)`
- `SetTraceData(string key, ISerializable value)` -> `SetTraceData(string key, string value)`

Modify the public `Callsite` APIs to return `strings` vs `ISerializables`.  The following APIs are impacted.  These are typically used by host integration for more advance manipulation of the trace data.
- `IList<ISerializable> GetOrphanedSerializables()` -> `IList<string> GetOrphanedSerializables()`
- `IList<ISerializable> GetAllSerializablesFromSingleRunTraceData` -> 
         `IList<string> GetAllSerializablesFromSingleRunTraceData`

Modify the public virtual method on `DynamoModel`.  I believe that DynamoRevit is the only downstream host that utilizes this mechanism.
-  `PostTraceReconciliation(Dictionary<Guid, List<ISerializable>> orphanedSerializables)` ->
         `PostTraceReconciliation(Dictionary<Guid, List<string>> orphanedSerializables)`

The second change to this PR is modification of the serialization mechanism used to store Element binding data in the file.  Although these are public, I don't know of any other consumer.  The primary change is modifying the `SingleRunTraceData` object main property `public ISerializable Data` -> `public string Data`. There are other methods on this class that change but all are `ISerializable` to `string`

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Trace mechanism has dropped support for insecure SOAP deserialization, and now requires clients to use strings. Data is still intended to be opaque for users.

### FYIs

@aparajit-pratap @mjkkirschner

### CODE REVIEW

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 1]
```diff
@@ -621,27 +436,36 @@ public string GetTraceDataToSave()
...
+        }
 
-                IFormatter formatter = new SoapFormatter();
-                formatter.Serialize(memoryStream, helper);
+        /// <summary>
+        /// Compress the input string via GZIP to Base64String
+        /// </summary>
+        /// <param name="json"></param>
+        /// <returns></returns>
+        public static string CompressSerializedTraceData(string json)
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/CallSite.cs:450]: do you anticipate host integrators needing this method?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 2]
```diff
@@ -2041,21 +1865,45 @@ public static StackValue PerformReturnTypeCoerce(FunctionEndPoint functionEndPoi
...
             RawTraceData callSiteData)
         {
-            var helper = TraceSerialiserHelper.FromCallSiteData(callSiteData.Data);
-            if (helper == null)
+            List<SingleRunTraceData> traceData = null;
+            try
+            {
+                //Optional Compression / Decompression
+                var data = DecompressSerializedTraceData(callSiteData.Data);
+                traceData = JsonConvert.DeserializeObject<List<SingleRunTraceData>>(data);
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1880]: perhaps it doesn't matter but I wonder if it might be more secure to use System.Text.Json - I think it has some different defaults when it comes to deserializing different types, I guess that could cause some issues for deserializing this class or other integrator specific trace classes...

https://github.com/DynamoDS/DynamoRevit/blob/master/src/Libraries/RevitServices/RevitTraceUtils.cs#L33

**Comment by mjkkirschner** [comment]
Two more thoughts
1. What happens if you load a graph with a lot of old format trace data- do we get tons of Json exceptions? Can we dump old data from v2 graphs instead?
2. I think there are some element binding performance graphs and tests floating around - I guess we’ll need up to update them as well.

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 1 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:450]: Probably not

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 2 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1880]: I think it is easy to swap out to System.Text.Json.  I tried it but required fully qualified namespace but works exactly the same.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1880]: Also we can update the RevitTraceUtils you reference above....  We will have to check with Vlad as he has removed that class in the NET7 D4R branch -> https://github.com/DynamoDS/DynamoRevit/blob/upgrade_to_net7/src/Libraries/RevitServices/RevitServices.csproj#L95

File: src/DynamoCore/Models/DynamoModel.cs [hunk 3]
```diff
@@ -1233,10 +1233,10 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
         {
             Debug.WriteLine("TRACE RECONCILIATION: {0} total serializables were orphaned.", obj.CallsiteToOrphanMap.SelectMany(kvp => kvp.Value).Count());
 
-            // The orphans will come back here as a dictionary of lists of ISerializables jeyed by their callsite id.
+            // The orphans will come back here as a dictionary of lists of strings jeyed by their callsite id.
```
Comment (aparajit-pratap) [inline src/DynamoCore/Models/DynamoModel.cs:1236]: keyed by their...

File: test/Engine/FFITarget/MinimalTraceTargets.cs [hunk 4]
```diff
@@ -118,34 +119,27 @@ public void Dispose()
...
 
 
     [Serializable]
-    internal class DummyDataHolder : ISerializable
+    internal class DummyDataHolder
     {
         public void GetObjectData(SerializationInfo info, StreamingContext context)
         {
             throw new NotImplementedException();
         }
```
Comment (aparajit-pratap) [inline test/Engine/FFITarget/MinimalTraceTargets.cs:140]: This can be removed.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 5]
```diff
@@ -463,14 +267,31 @@ public Guid CallSiteID
...
-            var helper = TraceSerialiserHelper.FromCallSiteData(serializedTraceData);
-            if (helper == null)
+            if (serializedTraceData == null)
             {
-                beforeFirstRunSerializables = new List<ISerializable>();
+                beforeFirstRunSerializables = new List<string>();
                 return;
             }
 
-            this.traceData = helper.TraceData;
```
Comment (saintentropy) [inline src/Engine/ProtoCore/Lang/CallSite.cs:473]: Maybe opportunity to pre-check and not do the try catch.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 6]
```diff
@@ -262,12 +262,12 @@ public Guid CallSiteID
...
 
         /// <summary>
-        /// Load the serialised data provided into this callsite's trace cache
+        /// Load the serialized data provided into this callsite's trace cache
         /// </summary>
         /// <param name="serializedTraceData">The data to load</param>
         public void LoadSerializedDataIntoTraceCache(string serializedTraceData)
         {
-            if (serializedTraceData == null)
+            if (serializedTraceData == null || CheckIfTraceDataIsLegacySOAPFormat(serializedTraceData))
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:270]: What happens if you skip the `CheckIfTraceDataIsLegacy...`?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 7]
```diff
@@ -562,6 +564,21 @@ private void UpdateCallsiteExecutionState(Object callsiteData, RuntimeCore runti
...
             }*/
         }
+        //TODO write a unit test for this.
+        private static bool CheckIfTraceDataIsLegacySOAPFormat(string base64EncodedTraceData)
+        {
+            var data = Convert.FromBase64String(base64EncodedTraceData);
+            if (data.Length > 71)
+            {
+                var header = Encoding.UTF8.GetString(data, 0, 72);
+                if (header == @"<SOAP-ENV:Envelope xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance""")
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:574]: Is it always 2001?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 8]
```diff
@@ -562,6 +564,21 @@ private void UpdateCallsiteExecutionState(Object callsiteData, RuntimeCore runti
                 }
             }*/
         }
+        //TODO write a unit test for this.
+        private static bool CheckIfTraceDataIsLegacySOAPFormat(string base64EncodedTraceData)
+        {
+            var data = Convert.FromBase64String(base64EncodedTraceData);
+            if (data.Length > 71)
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:571]: What are the first 71 chars? Does it include `<?xml version=”“1.0"” encoding=““utf-8"”?> ...`?

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 6 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:270]: then the try catch below will get hit and we'll end up doing a full base64 decode and catching an exception.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 9]
```diff
@@ -1892,7 +1909,7 @@ public static StackValue PerformReturnTypeCoerce(FunctionEndPoint functionEndPoi
             }
             catch
             {
-                //Do nothing.  Old Format data.
+                //Do nothing.
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1912]: I think it's good to catch them anyway, I could add some logs though if you think it's a good idea - we might hit unanticipated edge cases in deserialization with the new format.

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 7 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:574]: in all test data I saw - we could probably make the header even shorter / more tolerant.

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 8 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:571]: it's `<SOAP-ENV:Envelope xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"` exactly.

Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:571]: to clarify the reason I do this check is because if theres less than 71 bytes, the `GetString()` call below will fail.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 10]
```diff
@@ -458,19 +262,36 @@ public Guid CallSiteID
...
+            List<SingleRunTraceData> newTraceData = null;
+            try
+            {
+                //Optional Compression / Decompression
+                var decompressedTraceData = DecompressSerializedTraceData(serializedTraceData);
+                newTraceData = JsonConvert.DeserializeObject<List<SingleRunTraceData>>(decompressedTraceData);
+            }
+            catch
+            {
+                //do nothing.
```
Comment (pinzart90) [inline src/Engine/ProtoCore/Lang/CallSite.cs:285]: no logging needed here ?

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 7 above)
Comment (pinzart90) [reply src/Engine/ProtoCore/Lang/CallSite.cs:574]: Mayabe just check for `<SOAP-ENV:Envelope`..

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1912]: Yes, would be good to log into the Dynamo console.

**Comment by saintentropy** [comment]
Hey @mjkkirschner I also forgot we should remove the Soap Formatter 
Here -> https://github.com/DynamoDS/Dynamo/blob/master/src/Engine/ProtoCore/ProtoCore.csproj#L26

**Comment by saintentropy** [comment]
Here -> https://github.com/DynamoDS/Dynamo/blob/master/src/Engine/ProtoCore/Lang/CallSite.cs#L9
and here -> https://github.com/DynamoDS/Dynamo/blob/master/doc/distrib/License.rtf#L1667

File: src/NodeServices/TraceSupport.cs [hunk 11]
```diff
@@ -20,13 +20,13 @@ public static class TraceUtils
         internal const string __TEMP_REVIT_TRACE_ID = "{0459D869-0C72-447F-96D8-08A7FB92214B}-REVIT";
         // ReSharper restore InconsistentNaming
 
-        [ThreadStatic] private static Dictionary<String, ISerializable> _localStorageSlot; //= new Dictionary<string, ISerializable>();
```
Comment (saintentropy) [inline src/NodeServices/TraceSupport.cs:23]: Also I think we want to delete this class.

Comment (saintentropy) [reply src/NodeServices/TraceSupport.cs:23]: Oops not this one.

File: src/Engine/ProtoCore/Lang/TraceUtils.cs [hunk 12]
```diff
@@ -34,14 +34,14 @@ public static List<String> TEMP_GetTraceKeys()
         /// Returns a map of TraceID -> Objects
         /// </summary>
         /// <returns></returns>
-        public static Dictionary<String, ISerializable> GetObjectFromTLS()
```
Comment (saintentropy) [inline src/Engine/ProtoCore/Lang/TraceUtils.cs:37]: This class we can delete now.

Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/TraceUtils.cs:37]: this class is used in tests.

Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/TraceUtils.cs:37]: ah damn, looking at the wrong class.

Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/TraceUtils.cs:37]: done

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 13]
```diff
@@ -2041,21 +1886,50 @@ public static StackValue PerformReturnTypeCoerce(FunctionEndPoint functionEndPoi
...
 
         /// <summary>
-        /// Returns a flat collection of ISerializable objects from a serialized representation of a SingleRunTraceData object.
+        /// Returns a flat collection of strings from a serialized representation of a SingleRunTraceData object.
         /// </summary>
         /// <param name="callSiteData">The serialized representation of a SingleRunTraceData object.</param>
-        /// <returns>A flat collection of ISerializable objects.</returns>
-        public static IList<ISerializable> GetAllSerializablesFromSingleRunTraceData(
+        /// <returns>A flat collection of strings.</returns>
+        public static IList<string> GetAllSerializablesFromSingleRunTraceData(
```
Comment (saintentropy) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1893]: On thing we may be able to verify is does this need to be public.  Internal it is only called one place `GetOrphanedSerializablesAndClearHistoricalTraceData`...  Maybe in terms of quick validation of old data we can do it upstream and escape calling this one by one per node guid.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1893]: Hmmm. Looks like D4C3D calls it in their own version of `GetAllSerializablesFromSingleRunTraceData`.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1893]: So maybe stays public but we can also escape calls upstream and discuss with D4C3D the same optimization

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 6 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:269]: The thing we can do regarding your question about earlier catches and optimization of this check is handle something in here -> https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Graph/Workspaces/SerializationConverters.cs.  This is the only place dynamo core encounters the Soap data.  Maybe you can catch the old data there and then here the input string would be `null`.  That might also help you with your task @aparajit-pratap .  The nice thing is that the SerializationConverter is doing the whole `Bindings` section of the graph at once vs here were it is per `callsite`

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:269]: In the D4C3D case we may just need to communicate the same pattern as the load data from dwg file.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:269]: This optimization can be done later too.  This function could stay the same.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:269]: Actually even better, if you catch bad data at the file deserialization then even the `GetOrphanedSerializablesAndClearHistoricalTraceData` pass would be short circuited.  It is looking at `HomeWorkSpace.PreloadedTraceData` which is set during deserialization.

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:269]: Maybe it's just as simple as checking the DynamoVersion?   Not full proof but in theory you should never have XML graph or a 2.x graph with good data.

File: test/DynamoCoreTests/CallsiteTests.cs [hunk 14]
```diff
@@ -201,10 +196,10 @@ public void Callsite_ElementBinding_CustomNodes_multiple()
...
 
             BeginRun();
-            AssertPreviewValue("da39dbe5f59649b18c2fb6ca54acba7b", 111111);
-            AssertPreviewValue("2366239164a9441a8c4dcd981d9cf542", 222222);
-            AssertPreviewValue("8cfce012280342f3bd688520d68a7f66", 333333);
-            AssertPreviewValue("08448232ee094aad8280e9a99ed44f46", 444444);
+            AssertPreviewValue("da39dbe5f59649b18c2fb6ca54acba7b", 6);
+            AssertPreviewValue("2366239164a9441a8c4dcd981d9cf542", 7);
+            AssertPreviewValue("8cfce012280342f3bd688520d68a7f66", 10);
+            AssertPreviewValue("08448232ee094aad8280e9a99ed44f46", 11);
```
Comment (aparajit-pratap) [inline test/DynamoCoreTests/CallsiteTests.cs:202]: Sorry, what are these values that have changed?

Comment (mjkkirschner) [reply test/DynamoCoreTests/CallsiteTests.cs:202]: handwritten values were replace with values generated by interacting with the graph instead. It's a bit harder to write values directly into trace by hand since the data is bas64 encoded and gzipped
